clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Clash.Core.Evaluator.Models where
+
+import Control.Concurrent.Supply (Supply)
+import Control.Monad.State.Strict (State)
+import Control.DeepSeq (NFData(..))
+import Data.Bifunctor (first, second)
+import Data.Foldable (foldl')
+import Data.IntMap.Strict (IntMap)
+import qualified Data.IntMap.Strict as IntMap
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+
+import BasicTypes (InlineSpec(..))
+import SrcLoc (noSrcSpan)
+
+import Clash.Core.DataCon
+import Clash.Core.Literal
+import Clash.Core.Subst
+import Clash.Core.Term
+import Clash.Core.TyCon
+import Clash.Core.Type
+import Clash.Core.Var
+import Clash.Core.VarEnv
+import Clash.Driver.Types
+
+-- | An evaluator contains the basic functions that define partial evaluation.
+-- This consists of:
+--
+--   * evaluate, which evaluates terms to WHNF
+--   * quote, which recursively evaluates sub-terms and eta-expands to NF
+--
+-- Users of an evalautor can choose whether they convert terms to WHNF or NF,
+-- as the 'AsTerm' class (which converts back to Term) has instances for both.
+--
+data Evaluator = Evaluator
+  { evaluate       :: Term -> State Env Value
+  , quote          :: Value -> State Env Nf
+  }
+
+-- Local bindings are not stored in a VarEnv, as we still want to be
+-- able to access the keys (VarEnv simply indexes by unique).
+
+type EnvTyMap   = Map TyVar Type
+type EnvTmMap   = Map Id (Either Term Value)
+
+-- EnvGlobals refers to global bindings in scope during evalaution (i.e. other
+-- top-level functions used in a program).
+
+type EnvGlobals = VarEnv (Binding (Either Term Value))
+
+-- EnvPrimsIO refers to the result of IO operations performed in primitives
+-- during evalaution e.g. creating a new ByteArray in the GHC frontend.
+
+type EnvPrimsIO = (IntMap Value, Int)
+
+-- | An Environment contains all in scope terms and types while evaluating.
+-- This consists of
+--
+--   * local types and terms from the term being evaluated
+--   * top-level definitions in scope (i.e. global bindings)
+--   * IO results of primitive operations
+--   * in scope uniques and a means of generating uniques
+--
+-- When running the partial evaluator at different times, it may be desirable
+-- to keep the globals (some of which may have been partially evaluated) and
+-- the results of primitives in IO for future runs.
+--
+data Env = Env
+  { envTypes   :: EnvTyMap
+  , envLocals  :: EnvTmMap
+  , envGlobals :: EnvGlobals
+  , envPrimsIO :: EnvPrimsIO
+  , envTcMap   :: TyConMap
+  , envInScope :: InScopeSet
+  , envSupply  :: Supply
+  }
+
+instance Show Env where
+  show env = show (envTypes env, envLocals env, envPrimsIO env)
+
+instance NFData Env where
+  rnf env = rnf (envTypes env)
+    `seq` rnf (envLocals env)
+    `seq` rnf (envGlobals env)
+    `seq` rnf (envPrimsIO env)
```

Comment:
I think this either:

1. Needs a comment on why it breaks NFData's laws.
2. Should be removed and be auto-derived using Generics.

Reply:
Also: it looks like `Env` could benefit from strictness annotations on its fields.

Reply:
You can't derive `NFData Env` generically, as `Supply` doesn't have an instance IIRC.

But yes, I'll look at which fields made it fail and mention it

Reply:
`seq`ing Supply should be enough I think!

Reply:
Ah, so should I add

```haskell
 `seq` envInScope env `seq` envSupply env `seq ()
```

or something to the end?

Reply:
If you were planning on making it a strict data type anyway, just the `!` on the field should suffice. (Worth a comment though, might be confusing..)

I.e.,

```haskell

data Env = Env
  { envTypes   :: EnvTyMap
  , envLocals  :: EnvTmMap
  , envGlobals :: EnvGlobals
  , envPrimsIO :: EnvPrimsIO
  , envTcMap   :: TyConMap
  , envInScope :: InScopeSet
  , envSupply  :: !Supply
  }
```

(and for all the other fields you want to make strict, if any)

Reply:
The semantics in my sandbox branch are pretty strict anyway, so I see no reason to not make the data type strict too (at least in the heaps). I'm not sure if there would be any real benefit to strictifying the Supply / InScopeSet / TyConMap, that would need some profiling to investigate